services/motion/builtin/move_request.go

```diff
@@ -137,14 +137,14 @@ func (mr *moveRequest) AnchorGeoPose() *spatialmath.GeoPose {
 
 // execute attempts to follow a given Plan starting from the index percribed by waypointIndex.
 // Note that waypointIndex is an atomic int that is incremented in this function after each waypoint has been successfully reached.
-func (mr *moveRequest) execute(ctx context.Context, plan motionplan.Plan, waypointIndex *atomic.Int32) (state.ExecuteResponse, error) {
```

Comment:
Why is this changing?

Reply:
The motivation for this spawned from recognizing that `waypointIndex` should not be an argument for `obstaclesIntersectPlan` or `execute` due to timing issues.

Suppose we have constructed some plan with the following inputs `[ [0, 0, 0], [a, b, c], [d, e, f], [0, 0, 0] ]` and the following poses `[ startPose, middlePose, goalPose, goalPose ]`.

Suppose that when we enter `obstaclesIntersectPlan` our `wayPointIndex == 1`. Hence, our currentInputs are `[a, b, x]` such that `x < c`. We then query the vision service + camera for detections.. [ some amount of time passes ]

We now have gained our detections but suppose that our currentInputs are now `[d, e, y]` such that `y < f`.
Then, the wayPointIndex within `execute` would be updated to `2`. but the wayPointIndex known by `obstaclesIntersectPlan` would be 1. This would result in an incorrect interpolation of the plan.

Reply:
Yes, I agree that this fixing a latent bug.

Reply:
I don't think this actually fixed the race, but rather simply moved it elsewhere; see comment here:

https://github.com/viamrobotics/rdk/pull/3718/files#r1535927476

Reply:
I will fix the race, thank you for noticing that it was not correctly handled.